controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +58,787 @@ ControllerManager::load_controller(
+    RCLCPP_INFO(get_logger(), "Controller manager: Stopping all running controllers");
+    std::vector<std::string> empty;
+    if (switch_controller(
+        empty, running_controllers,
+        controller_manager_msgs::srv::SwitchController::Request::BEST_EFFORT) !=
+      controller_interface::return_type::SUCCESS)
+    {
+      RCLCPP_ERROR(
+        get_logger(),
+        "Controller manager: Cannot reload controller libraries because failed to stop "
+        "running controllers");
+      response->ok = false;
+      return;
+    }
+    for (const auto & controller : loaded_controllers) {
+      if (unload_controller(controller) != controller_interface::return_type::SUCCESS) {
+        RCLCPP_ERROR(
+          get_logger(), "Controller manager: Cannot reload controller libraries because "
+          "failed to unload controller '%s'",
+          controller.c_str());
+        response->ok = false;
+        return;
+      }
+    }
+    get_controller_names(loaded_controllers);
+  }
+  assert(loaded_controllers.empty());
+
+  // Force a reload on all the PluginLoaders (internally, this recreates the plugin loaders)
+  for (const auto & controller_loader : loaders_) {
+    controller_loader->reload();
+    RCLCPP_INFO(
+      get_logger(), "Controller manager: reloaded controller libraries for '%s'",
+      controller_loader->getName().c_str());
+  }
+
+  response->ok = true;
+
+  RCLCPP_DEBUG(get_logger(), "reload libraries service finished");
+}
+
+void ControllerManager::switch_controller_service_cb(
+  const std::shared_ptr<controller_manager_msgs::srv::SwitchController::Request> request,
+  std::shared_ptr<controller_manager_msgs::srv::SwitchController::Response> response)
+{
+  // lock services
+  RCLCPP_DEBUG(get_logger(), "switching service called");
+  std::lock_guard<std::mutex> guard(services_lock_);
+  RCLCPP_DEBUG(get_logger(), "switching service locked");
+
+  response->ok = switch_controller(
+    request->start_controllers, request->stop_controllers, request->strictness,
+    request->start_asap, request->timeout) == controller_interface::return_type::SUCCESS;
+
+  RCLCPP_DEBUG(get_logger(), "switching service finished");
+}
+
+void ControllerManager::unload_controller_service_cb(
+  const std::shared_ptr<controller_manager_msgs::srv::UnloadController::Request> request,
+  std::shared_ptr<controller_manager_msgs::srv::UnloadController::Response> response)
+{
+  // lock services
+  RCLCPP_DEBUG(
+    get_logger(), "unloading service called for controller '%s' ",
+    request->name.c_str());
+  std::lock_guard<std::mutex> guard(services_lock_);
+  RCLCPP_DEBUG(get_logger(), "unloading service locked");
+
+  response->ok = unload_controller(request->name) == controller_interface::return_type::SUCCESS;
+
+  RCLCPP_DEBUG(
+    get_logger(), "unloading service finished for controller '%s' ",
+    request->name.c_str());
+}
+
+void ControllerManager::get_controller_names(std::vector<std::string> & names)
```

Comment:
why not returning a vector here? RVO should kick in quite nicely here.

Reply:
Will do.

Reply:
good idea, big fan of RVO

Reply:
done